File: 2023/day02/app/Main.hs
import Control.Concurrent (waitQSem)
import Data.List (isPrefixOf, maximumBy)
import Data.List.Split (splitOn)

main :: IO ()
main = do
  part1
  part2

parse :: String -> [(Int, [(Int, Int, Int)])]
parse contents = do
  let linesOfFile = lines contents
  [let split = splitID x in (fst split, parseSets (snd split)) | x <- linesOfFile]

splitID :: String -> (Int, String)
splitID game = (read (drop 5 (head (splitOn ": " game))), last (splitOn ": " game))

parseSets :: String -> [(Int, Int, Int)]
parseSets sets = map parseSet (splitOn "; " sets)

parseSet :: String -> (Int, Int, Int)
parseSet set =
  (fst (parse "red"), fst (parse "green"), fst (parse "blue"))
  where
    parse colour = case filter (filterColour colour) (map parseColour (splitOn ", " set)) of
      [] -> (0, colour)
      x -> head x

parseColour :: String -> (Int, String)
parseColour string =
  (read (head cube), last cube)
  where
    cube = words string

filterColour :: String -> (Int, String) -> Bool
filterColour colour str = snd str == colour

part1 :: IO ()
part1 = do
  contents <- readFile "inputs/main.input"
  let games = parse contents
  let test = filter (all hasEnough . snd) games
  print (sum (map fst test))

hasEnough :: (Int, Int, Int) -> Bool
hasEnough (red, green, blue) = red <= 12 && green <= 13 && blue <= 14

part2 :: IO ()
part2 = do
  contents <- readFile "inputs/main.input"
  let games = parse contents
  let maxes = map multiplyMarbles games
  print (sum maxes)

multiplyMarbles :: (Int, [(Int, Int, Int)]) -> Int
multiplyMarbles (_, xs) = maximum (map red xs) * maximum (map green xs) * maximum (map blue xs)
  where
    red (r, g, b) = r
    green (r, g, b) = g
    blue (r, g, b) = b

File: 2023/day02/src/main.rs
use std::error::Error;

use advent_of_code_lib::{self, Solver};

fn main() -> Result<(), Box<dyn Error>> {
    advent_of_code_lib::run_and_print(Day, "2023", "2")
}

struct Day;

impl Solver for Day {
    fn part1(&self, input: &str) -> String {
        input
            .lines()
            .filter_map(|f| {
                let mut split = f.split(": ");
                let id = split
                    .next()
                    .unwrap()
                    .replace("Game ", "")
                    .parse::<u32>()
                    .unwrap();

                let sets = split.next().unwrap().split("; ");
                for st in sets {
                    let balls = st.split(", ");
                    for ball in balls {
                        let mut comp = ball.split(" ");
                        let num = comp.next().unwrap().parse::<u32>().unwrap();
                        let colour = comp.next().unwrap();

                        if colour == "red" && num > 12 {
                            return None;
                        }
                        if colour == "green" && num > 13 {
                            return None;
                        }
                        if colour == "blue" && num > 14 {
                            return None;
                        }
                    }
                }
                return Some(id);
            })
            .sum::<u32>()
            .to_string()
    }

    fn part2(&self, input: &str) -> String {
        input
            .lines()
            .filter_map(|f| {
                let mut split = f.split(": ");
                split
                    .next()
                    .unwrap()
                    .replace("Game ", "")
                    .parse::<u32>()
                    .unwrap();

                let sets = split.next().unwrap().split("; ");
                let (red, green, blue) = sets.fold((0, 0, 0), |(red, green, blue), st| {
                    let balls = st.split(", ");
                    let (r, g, b) =
                        balls.fold((0, 0, 0), |(mut red, mut green, mut blue), ball| {
                            let mut comp = ball.split(" ");
                            let num = comp.next().unwrap().parse::<u32>().unwrap();
                            let colour = comp.next().unwrap();

                            match colour {
                                "red" => red = red.max(num),
                                "green" => green = green.max(num),
                                "blue" => blue = blue.max(num),
                                _ => panic!("Invalid state"),
                            }
                            return (red, green, blue);
                        });

                    return (red.max(r), green.max(g), blue.max(b));
                });
                return Some(red * green * blue);
            })
            .sum::<u32>()
            .to_string()
    }
}
